### PR TITLE
MINIFICPP-1202 - Handle C2 requests/responses using MinifiConcurrentQueue

### PR DESCRIPTION
**Pull request by hunyadi-dev**
Thank you for submitting a contribution to Apache NiFi - MiNiFi C++.

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [ ] Is there a JIRA ticket associated with this PR? Is it referenced
     in the commit message?

- [ ] Does your PR title start with MINIFICPP-XXXX where XXXX is the JIRA number you are trying to resolve? Pay particular attention to the hyphen "-" character.

- [ ] Has your PR been rebased against the latest commit within the target branch (typically master)?

- [ ] Is your initial contribution a single, squashed commit?

### For code changes:
- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)?
- [ ] If applicable, have you updated the LICENSE file?
- [ ] If applicable, have you updated the NOTICE file?

### For documentation related changes:
- [ ] Have you ensured that format looks appropriate for the output in which it is rendered?

### Note:
Please ensure that once the PR is submitted, you check travis-ci for build issues and submit an update to your PR as soon as possible.

### CODE REVIEW

**Comment by hunyadi-dev** [comment]
This PR is dependent on #773. Locking until that one is merged.

File: libminifi/src/c2/C2Agent.cpp [hunk 1]
```diff
@@ -75,54 +78,54 @@ C2Agent::C2Agent(const std::shared_ptr<core::controller::ControllerServiceProvid
...
-            catch(...) {
-              logger_->log_error("Unknonwn exception occurred while consuming payload.");
-            }
-          }while(!requests.empty() && ++count < max_c2_responses);
+    if (protocol_.load() != nullptr) {
+      std::vector<C2Payload> payload_batch;
+      payload_batch.reserve(max_c2_responses);
+      auto getRequestPayload = [&payload_batch] (C2Payload&& payload) { payload_batch.emplace_back(std::move(payload)); };
+      for (std::size_t attempt_num = 0; attempt_num < max_c2_responses; ++attempt_num) {
+        if (!requests.consumeWaitFor(getRequestPayload, std::chrono::seconds(1))) {
```
Comment (arpadboda) [inline libminifi/src/c2/C2Agent.cpp:86]: I don't think this behaviour is identical. 
The original waited at most 1 sec to acquire the mutex and grabbed as many events as it could.

Now we grab as many events as we can, waiting at most 1s for each.

File: libminifi/src/c2/C2Agent.cpp [hunk 2]
```diff
@@ -75,54 +78,54 @@ C2Agent::C2Agent(const std::shared_ptr<core::controller::ControllerServiceProvid
...
-      {
-        std::lock_guard<std::timed_mutex> lock(queue_mutex, std::adopt_lock);
-        if (responses.empty()) {
-          return utils::TaskRescheduleInfo::RetryIn(std::chrono::milliseconds(C2RESPONSE_POLL_MS));
-        }
-        payload = std::move(responses.back());
-        responses.pop_back();
+  c2_consumer_ = [&] {
+    if (responses.size()) {
+      if (!responses.consumeWaitFor([this](C2Payload&& e) { extractPayload(std::move(e)); }, std::chrono::seconds(1))) {
```
Comment (arpadboda) [inline libminifi/src/c2/C2Agent.cpp:126]: Here as well.

File: libminifi/src/c2/C2Agent.cpp (hunk 1 above)
Comment (hunyadi-dev) [reply libminifi/src/c2/C2Agent.cpp:86]: Will add a `wait_until` method to the queue and max out on 1 sec total time here.

File: libminifi/src/c2/C2Agent.cpp (hunk 2 above)
Comment (hunyadi-dev) [reply libminifi/src/c2/C2Agent.cpp:127]: This is the same as before, right?

File: libminifi/src/c2/C2Agent.cpp (hunk 1 above)
Comment (hunyadi-dev) [reply libminifi/src/c2/C2Agent.cpp]: Added `wait_until`.

File: libminifi/src/c2/C2Agent.cpp [hunk 3]
```diff
@@ -75,54 +78,55 @@ C2Agent::C2Agent(const std::shared_ptr<core::controller::ControllerServiceProvid
...
-      checkTriggers();
+        try {
+          performHeartBeat();
+        }
+        catch (const std::exception &e) {
+          logger_->log_error("Exception occurred while performing heartbeat. error: %s", e.what());
+        }
+        catch (...) {
+          logger_->log_error("Unknonwn exception occurred while performing heartbeat.");
+        }
```
Comment (szaszm) [inline libminifi/src/c2/C2Agent.cpp:115]: incorrect/misleading indentation

File: libminifi/src/c2/C2Agent.cpp [hunk 4]
```diff
@@ -75,54 +78,55 @@ C2Agent::C2Agent(const std::shared_ptr<core::controller::ControllerServiceProvid
...
-              logger_->log_error("Unknonwn exception occurred while consuming payload.");
-            }
-          }while(!requests.empty() && ++count < max_c2_responses);
+    if (protocol_.load() != nullptr) {
+      std::vector<C2Payload> payload_batch;
+      payload_batch.reserve(max_c2_responses);
+      auto getRequestPayload = [&payload_batch] (C2Payload&& payload) { payload_batch.emplace_back(std::move(payload)); };
+      const std::chrono::system_clock::time_point timeout_point = std::chrono::system_clock::now() + std::chrono::milliseconds(1);
+      for (std::size_t attempt_num = 0; attempt_num < max_c2_responses; ++attempt_num) {
+        if (!requests.consumeWaitUntil(getRequestPayload, timeout_point)) {
```
Comment (szaszm) [inline libminifi/src/c2/C2Agent.cpp:87]: Previously we took the immediately available payloads only. Now we wait 1ms. Why is this change?

If we reverted to the old way, it would also no longer be a problem to reunify the first two loops and eliminate the need for the temporary buffer (`payload_batch`).

Comment (hunyadi-dev) [reply libminifi/src/c2/C2Agent.cpp:87]: Good catch, we want to wait 1 second, not 1 millisecond on locking.

File: libminifi/src/c2/C2Agent.cpp (hunk 3 above)
Comment (hunyadi-dev) [reply libminifi/src/c2/C2Agent.cpp:115]: That is correct, that is one-too-many indents due to the `std::for_each` actually taking up two identation levels.

File: libminifi/src/c2/C2Agent.cpp [hunk 5]
```diff
@@ -75,54 +78,55 @@ C2Agent::C2Agent(const std::shared_ptr<core::controller::ControllerServiceProvid
...
-      {
-        std::lock_guard<std::timed_mutex> lock(queue_mutex, std::adopt_lock);
-        if (responses.empty()) {
-          return utils::TaskRescheduleInfo::RetryIn(std::chrono::milliseconds(C2RESPONSE_POLL_MS));
-        }
-        payload = std::move(responses.back());
-        responses.pop_back();
+  c2_consumer_ = [&] {
+    if (responses.size()) {
+      if (!responses.consumeWaitFor([this](C2Payload&& e) { extractPayload(std::move(e)); }, std::chrono::seconds(1))) {
```
[the remark this answers is not in the document]
Comment (hunyadi-dev) [reply libminifi/src/c2/C2Agent.cpp:127]: As far as I can see, this is the same behaviour:

- If no elements were present in the queue, we did not call `extractPayload()` at all.
- If there were elements, but by the time we wanted to grab the lock the queue became busy, we waited 1 second and tried fetching an element
  - If by this time we succeeded, we called `extractPayload()` with this dequeued data
  - Otherwise we used the default constructed payload: `C2Payload{ Operation::HEARTBEAT }` instead.

To me `!empty()` feels like something that is easily mistaken for `empty()`, `size()` is guaranteed to be constant time for `std::deque` and `empty()` is probably implemented as `size() == 0` anyway, right?

File: libminifi/include/c2/C2Agent.h [hunk 6]
```diff
@@ -41,8 +41,15 @@ namespace org {
 namespace apache {
 namespace nifi {
 namespace minifi {
+
+namespace utils {
+template <typename T>
+class ConditionConcurrentQueue;
+}
```
[the remark this answers is not in the document]
Comment (hunyadi-dev) [reply libminifi/include/c2/C2Agent.h:48]: You are completely right, at first I thought we would only need the declaration and left this here. Will replace it with a proper include.

File: libminifi/src/c2/C2Agent.cpp (hunk 4 above)
Comment (szaszm) [reply libminifi/src/c2/C2Agent.cpp:87]: The mutex inside the queue is not `timed_mutex` so we wait indefinitely for locking and only wait definitely for content.

File: libminifi/src/c2/C2Agent.cpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (szaszm) [reply libminifi/src/c2/C2Agent.cpp:127]: 1. My preference is `!empty()`, because it reads as "not empty", which is the intention here. On the other hand I'm ok with using `size() > 0`, as that's only one logical step away from the intention, but not plain `size()` because that implicit int -> bool conversion is surprising to the reader, like every implicit conversion that converts to something that's not logically "the same thing". I think the readers mind goes like this:
    1. if the size of responses. wtf, that doesn't make sense
    2. Ah, implicit int -> bool conversion, so "if the size of the responses is not zero".
    3. That's "if the responses are not empty", i.e. "if there are responses"

    Sorry for the long comment, I hope this makes sense.

2. Sorry for the accusation, you're right about the second point. I must have been salty because of something when writing this. :(

    However, this raises another problem, which is that I, as the reader, misunderstood the code, so it's probably too complex. I suggest extracting the lambda and maybe even the consume call, so that the identifiers can guide the reader.

3. \-

```suggestion
    if (responses.size() > 0) {  // or (!responses.empty()), or (not responses.empty())
      const auto on_dequeue = [this](C2Payload&& payload) { extractPayload(std::move(payload)); };
      const auto dequeue_success = responses.consumeWaitFor(on_dequeue, std::chrono::seconds{1});
      if (!dequeue_success) {
```

[the remark this answers is not in the document]
Comment (hunyadi-dev) [reply libminifi/src/c2/C2Agent.cpp:127]: 1.
> Call empty instead of checking size() against zero. 

Of course I would do that, my problem is with the hidden bangs when reading code :) Unfortunately, the current linter setting warns on `! empty()` which would make it at least stand out. Maybe it is just personal because I have committed incorrect code due to not noticing a `!`, but always felt reading `if(vector.size())` as `if vector has elements`.

tldr; will change it

2.
Will extract the logic out.

3.
I dont want to commit that comment on, can I somehow edit the suggestion? :)

[the remark this answers is not in the document]
Comment (szaszm) [reply libminifi/src/c2/C2Agent.cpp:127]: I don't think you can edit my suggestion (in my previous comment), but you can manually apply the desired changes in your working directory, or I can edit my comment for you if you specify what changes do you want.
Do you want to keep the current suggestion except the comment, or change something else, too?

File: libminifi/src/c2/C2Agent.cpp (hunk 3 above)
Comment (szaszm) [reply libminifi/src/c2/C2Agent.cpp:115]: The second set of try/catch blocks are outside of the `for_each` lambda

File: libminifi/include/c2/C2Agent.h (hunk 6 above)
[the remark this answers is not in the document]
Comment (hunyadi-dev) [reply libminifi/include/c2/C2Agent.h]: Replaced with include.

File: libminifi/src/c2/C2Agent.cpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (hunyadi-dev) [reply libminifi/src/c2/C2Agent.cpp]: Corrected as requested.

File: libminifi/src/c2/C2Agent.cpp (hunk 4 above)
Comment (hunyadi-dev) [reply libminifi/src/c2/C2Agent.cpp]: What is the recommendation here?

Comment (szaszm) [reply libminifi/src/c2/C2Agent.cpp]: I understand that the indefinite waiting on the mutex is because of the use of `MinifiConcurrentQueue`. But why did we introduce a waiting on the queue condition variable for elements?

I just want an explanation, to see that the change was intentional and why.

Comment (hunyadi-dev) [reply libminifi/src/c2/C2Agent.cpp]: I just thought that if one second waiting time was reasonable here, than I should keep this if-it-works-in-one-second-it-is-good behaviour. Maybe there should be a timeout on the mutex itself, but I wonder what solution sounds reasonable here?

Comment (szaszm) [reply libminifi/src/c2/C2Agent.cpp]: In the old ideal (and likely common) case before, we got the mutex instantly, took whatever was in the queue and continued execution.

In the new common case, we lock the mutex, take whatever is in the queue, wait 1 second for more to arrive and continue execution.

The common case execution time is increased by almost 1 second, which is probably orders of magnitude more than how much time is spent executing business logic. I'd like to get rid of that waiting if it's not needed. I suggest using `consume` from `ConcurrentQueue` without waiting on the condition variable. This way if we assume little waiting for the mutex, the execution time stays similar.

**Comment by szaszm** [comment]
https://github.com/apache/nifi-minifi-cpp/pull/776#discussion_r431809781

As described here, the average producer execution time has increased to 1 second and I think also the heartbeat interval is effectively restricted to at most once a second in absence of other C2 requests, but I didn't test this observation. 

Until this is fixed (or explained why my observation is wrong) I can't +1 this PR.

**Comment by arpadboda** [comment]
> [#776 (comment)](https://github.com/apache/nifi-minifi-cpp/pull/776#discussion_r431809781)
> 
> As described here, the average producer execution time has increased to 1 second and I think also the heartbeat interval is effectively restricted to at most once a second in absence of other C2 requests, but I didn't test this observation.
> 
> Until this is fixed (or explained why my observation is wrong) I can't +1 this PR.

Good point, I agree that we should simply process whatever is there without waiting. 

The wait_until interface of the concurrentqueue is still useful, so I would keep that, just don't use in the particular case.
